modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/InboundConnectionHandler.java

```diff
@@ -0,0 +1,930 @@
+            }
+            finally {
+                connectGate.leave();
+            }
+        }
+        else {
+            Object consistentId = ses.meta(CONSISTENT_ID_META);
+
+            assert consistentId != null;
+
+            if (msg instanceof RecoveryLastReceivedMessage) {
+                metricsLsnr.onMessageReceived(msg, consistentId);
+
+                GridNioRecoveryDescriptor recovery = ses.outRecoveryDescriptor();
+
+                if (recovery != null) {
+                    RecoveryLastReceivedMessage msg0 = (RecoveryLastReceivedMessage)msg;
+
+                    if (log.isDebugEnabled()) {
+                        log.debug("Received recovery acknowledgement [rmtNode=" + connKey.nodeId() +
+                            ", connIdx=" + connKey.connectionIndex() +
+                            ", rcvCnt=" + msg0.received() + ']');
+                    }
+
+                    recovery.ackReceived(msg0.received());
+                }
+
+                return;
+            }
+            else {
+                GridNioRecoveryDescriptor recovery = ses.inRecoveryDescriptor();
+
+                if (recovery != null) {
+                    long rcvCnt = recovery.onReceived();
+
+                    if (rcvCnt % cfg.ackSendThreshold() == 0) {
+                        if (log.isDebugEnabled()) {
+                            log.debug("Send recovery acknowledgement [rmtNode=" + connKey.nodeId() +
+                                ", connIdx=" + connKey.connectionIndex() +
+                                ", rcvCnt=" + rcvCnt + ']');
+                        }
+
+                        ses.systemMessage(new RecoveryLastReceivedMessage(rcvCnt));
+
+                        recovery.lastAcknowledged(rcvCnt);
+                    }
+                }
+                else if (connKey.dummy()) {
+                    assert msg instanceof NodeIdMessage : msg;
+
+                    TcpCommunicationNodeConnectionCheckFuture fut = ses.meta(SES_FUT_META);
+
+                    assert fut != null : msg;
+
+                    fut.onConnected(U.bytesToUuid(((NodeIdMessage)msg).nodeIdBytes(), 0));
+
+                    nioSrvWrapper.nio().closeFromWorkerThread(ses);
+
+                    return;
+                }
+            }
+
+            metricsLsnr.onMessageReceived(msg, consistentId);
+
+            IgniteRunnable c;
+
+            if (cfg.messageQueueLimit() > 0) {
+                GridNioMessageTracker tracker = ses.meta(TRACKER_META);
+
+                if (tracker == null) {
+                    GridNioMessageTracker old = ses.addMeta(TRACKER_META, tracker =
+                        new GridNioMessageTracker(ses, cfg.messageQueueLimit()));
+
+                    assert old == null;
+                }
+
+                tracker.onMessageReceived();
+
+                c = tracker;
+            }
+            else
+                c = NOOP;
+
+            lsnr.onMessage(connKey.nodeId(), msg, c);
```

Comment:
Can 'lsnr' be null and cause NPE?
In 'onDisconnected' method we expect a nullable 'lsnr'.